pandas/core/frame.py

```diff
@@ -7022,56 +7022,154 @@ def cov(self, min_periods=None):
         other : DataFrame, Series
         axis : {0 or 'index', 1 or 'columns'}, default 0
             0 or 'index' to compute column-wise, 1 or 'columns' for row-wise
+        how : {'pairwise', 'all'}, default 'pairwise' if other is a DataFrame
+            and default 'all' if other is a Series object
+            * 'pairwise' to compute correlation between rows or columns with
+              the same name/index,
+            * 'all' to compute correlation between all possible pairs of rows
+              or columns
+
+            .. versionadded:: 0.24.0
         drop : boolean, default False
             Drop missing indices from result, default returns union of all
+        method : {'pearson', 'kendall', 'spearman'} or callable
+            * 'pearson' : standard correlation coefficient
+            * 'kendall' : Kendall Tau correlation coefficient
+            * 'spearman' : Spearman rank correlation
+            * callable : callable with input two 1d ndarrays
+                 and returning a float
+
+            .. versionadded:: 0.24.0
+        min_periods : int, optional
+            Minimum number of observations required per pair of columns
+            to have a valid result. Currently only available for pearson
+            and spearman correlation
+
+            .. versionadded:: 0.24.0
 
         Returns
         -------
-        correls : Series
+        correl : Series (how='pairwise') or DataFrame (how='all')
         """
+
         axis = self._get_axis_number(axis)
         this = self._get_numeric_data()
 
         if isinstance(other, Series):
-            return this.apply(other.corr, axis=axis)
+            return this.apply(other.corr, axis=axis,
+                              method=method, min_periods=min_periods)
 
-        other = other._get_numeric_data()
+        if not isinstance(other, DataFrame):
+            raise TypeError("'other' parameter should be a DataFrame "
+                            "or a Series object")
 
-        left, right = this.align(other, join='inner', copy=False)
+        other = other._get_numeric_data()
 
-        # mask missing values
-        left = left + right * 0
-        right = right + left * 0
+        if isinstance(this.columns, MultiIndex) \
+                or isinstance(other.columns, MultiIndex):
+            raise ValueError("MultiIndex is not supported")
 
         if axis == 1:
-            left = left.T
-            right = right.T
+            this = this.transpose()
+            other = other.transpose()
+
+        if (len(set(this.columns)) != len(this.columns)) \
+                or (len(set(other.columns)) != len(other.columns)):
+            raise ValueError("Non-unique columns are not supported")
+
+        if how == 'all':
+            corr = np.zeros((this.shape[1], other.shape[1]))
+
+            for i in range(len(this.columns)):
+                for j in range(len(other.columns)):
+                    corr[i, j] = this.iloc[:, i].corr(
```

Comment:
is this double computing? e.g. is [i, j] == [j, i]?

Reply:
No — It's not a square matrix. Columns != Rows.

Reply:
add a comment explaining this